src/custom-element.js

```diff
@@ -987,9 +977,17 @@ function createBaseCustomElementClass(win) {
       }
       if (!this.ampdoc_) {
         // Ampdoc can now be initialized.
-        const ampdocService = Services.ampdocServiceFor(
-            this.ownerDocument.defaultView);
-        this.ampdoc_ = ampdocService.getAmpDoc(this);
+        const win = this.ownerDocument.defaultView;
+        const ampdocService = Services.ampdocServiceFor(win);
+        const ampdoc = ampdocService.getAmpDoc(this);
+        this.ampdoc_ = ampdoc;
+        // Load the pre-stubbed extension if needed.
+        const extensionId = this.tagName.toLowerCase();
+        if (isStub(this.implementation_) &&
+            !ampdoc.declaresExtension(extensionId)) {
+          Services.extensionsFor(win).installExtensionForDoc(
```

Comment:
This case only covers stub redirects and only for legacy builtins (e.g. `amp-video`) that became proper extensions. This is direct substitute of this [code](https://github.com/ampproject/amphtml/blob/master/src/element-stub.js#L35). But since I can't access `ampdoc` in the stub element's constructor, it was more convenient to move it here.